hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/supervisor/process/ConsumerProcessSupervisorSignalsTest.groovy

```diff
@@ -0,0 +1,160 @@
+    HermesMetrics hermesMetrics
+
+    def setup() {
+        configFactory.overrideProperty(Configs.CONSUMER_BACKGROUND_SUPERVISOR_KILL_AFTER, 100)
+        hermesMetrics = new HermesMetrics(new MetricRegistry(), new PathsCompiler("localhost"))
+
+        consumerProcessSupervisor = new ConsumerProcessSupervisor(
+                new ConsumersExecutorService(configFactory, hermesMetrics),
+                retransmitter,
+                new Clock.SystemClock(),
+                hermesMetrics,
+                configFactory,
+                consumerFactory)
+    }
+
+    def cleanup() {
+        consumerProcessSupervisor.shutdown()
+    }
+
+    def "should restart unhealthy process"() {
+        given:
+        def consumer1 = Stub(Consumer)
+        def consumer2 = Stub(Consumer)
+        def consumers = [consumer1, consumer2]
+
+        def conditions = new PollingConditions(timeout: 2)
+        def createConsumerFactoryMethodCallsCount = 0
+
+        consumer1.consume(_) >> {
+            Runnable signalsInterrupt -> signalsInterrupt.run()
+        }
+        consumer2.consume(_) >> {
+            Runnable signalsInterrupt -> signalsInterrupt.run()
+        }
+
+        consumerFactory.createConsumer(subscription) >>
+                { subscription -> consumers[createConsumerFactoryMethodCallsCount++]}
+
+        consumerProcessSupervisor.accept(
+                Signal.of(Signal.SignalType.START, subscription.getQualifiedName(), subscription))
+        consumerProcessSupervisor.run()
+
+        when:
+        consumerProcessSupervisor.accept(
+                Signal.of(Signal.SignalType.RESTART_UNHEALTHY, subscription.getQualifiedName()))
+
+        then:
+        conditions.eventually {
+            consumerProcessSupervisor.run()
+            assert createConsumerFactoryMethodCallsCount == 2
+            assert consumerProcessSupervisor.countRunningProcesses() == 1
+        }
+    }
+
+    @Unroll
+    def "should #signalName process"() {
+        given:
+        def delay = 0.5
+        def timestampContainer = new TimestampContainer()
+
+        def consumer = Stub(Consumer)
+        def conditions = new PollingConditions(timeout: 4 * delay, delay: delay)
+
+        consumer.consume(_) >> { Runnable signalsInterrupt ->
+            signalsInterrupt.run()
+            timestampContainer.update()
+        }
+
+        consumerFactory.createConsumer(subscription) >> consumer
+        consumerProcessSupervisor.accept(
+                Signal.of(Signal.SignalType.START, subscription.getQualifiedName(), subscription))
+        consumerProcessSupervisor.run()
+
+        when:
+        consumerProcessSupervisor.accept(
+                Signal.of(signalType, subscription.getQualifiedName(), subscription))
+        consumerProcessSupervisor.run()
+
+        then:
+        conditions.within(4 * delay) {
+            assert System.currentTimeMillis() - timestampContainer.timestamp > 2 * delay * 1000
```

Comment:
This is quite tricky, we just measure diff time between time in condition thread and thread in consumer process. If consumer process is stopped/killed it does not update its timestamp and diff exceeds some threshold.

Reply:
Ok, but why `2 *` and what's delay (and why it is `0.5`). I assume `* 1000` is for converting to seconds.

Reply:
Ok, it maybe a bit confusing. Firstly "delay" is passed to PollingConditions and it basically means the rate (in seconds) at which condition will be checked. So here we've got every 0.5 seconds. 
```java
conditions.within(4 * delay)
```
This line of code means that we can spare at most 2 seconds for waiting for signal to be processed (I just think it is reasonable max time - maybe shorter would be ok too, signals are on the queue and there is supervisor run interval set on 100ms so it tries to process signals at most 20 times)
```java
assert System.currentTimeMillis() - timestampContainer.timestamp > 2 * delay * 1000
```
here 2 * delay * 1000 means that the condition loop ran at least 2 times  and the timestampContainer.timestamp was not updated